chainstate/src/detail/tests/normal_tokens.rs

```diff
@@ -0,0 +1,867 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::anyonecanspend_address;
+use crate::detail::tests::test_framework::TestFramework;
+use crate::detail::transaction_verifier::error::ConnectTransactionError;
+use crate::detail::CheckBlockTransactionsError;
+use crate::{
+    detail::{tests::TestBlockInfo, CheckBlockError, TokensError},
+    BlockError, BlockSource,
+};
+use chainstate_types::BlockIndex;
+use common::chain::block::BlockReward;
+use common::chain::config::TOKEN_MIN_ISSUANCE_FEE;
+use common::{
+    chain::{
+        block::{timestamp::BlockTimestamp, Block, ConsensusData},
+        signature::inputsig::InputWitness,
+        tokens::{token_id, OutputValue, TokenData, TokenId},
+        GenBlock, OutPointSourceId, OutputPurpose, Transaction, TxInput, TxOutput,
+    },
+    primitives::{time, Amount, Id, Idable},
+};
+use std::vec;
+
+enum ParentBlock {
+    BestBlock,
+    BlockId(Id<GenBlock>),
+}
+
+fn process_token(
+    test_framework: &mut TestFramework,
+    parent_block: ParentBlock,
+    values: Vec<OutputValue>,
+) -> Result<Option<BlockIndex>, BlockError> {
+    process_token_ex(test_framework, vec![], parent_block, values).map(|(_block_id, result)| result)
+}
+
+fn is_token_burned(output_value: &OutputValue) -> bool {
+    match output_value {
+        OutputValue::Coin(_) => false,
+        OutputValue::Token(token_data) => match token_data {
+            TokenData::TokenTransferV1 {
+                token_id: _,
+                amount: _,
+            } => false,
+            TokenData::TokenIssuanceV1 {
+                token_ticker: _,
+                amount_to_issue: _,
+                number_of_decimals: _,
+                metadata_uri: _,
+            } => false,
+            TokenData::TokenBurnV1 {
+                token_id: _,
+                amount_to_burn: _,
+            } => true,
+        },
+    }
+}
+
+fn process_token_ex(
+    test_framework: &mut TestFramework,
+    additional_inputs: Vec<TxInput>,
+    parent_block: ParentBlock,
+    values: Vec<OutputValue>,
+) -> Result<(Block, Option<BlockIndex>), BlockError> {
+    let receiver = anyonecanspend_address();
+    let parent_block_id = match parent_block {
```

Comment:
Why can't you just pass the best block from the test_framework? Isn't it just a function call there? Why the enum? This is just more complexity with no real benefit.

Reply:
I think, we can resolve this because now I use the builder pattern there